core/src/main/java/org/infinispan/interceptors/locking/ClusteringDependentLogic.java

```diff
@@ -508,47 +584,79 @@ protected void commitSingleEntry(CacheEntry entry, Metadata metadata, FlagAffect
                         builder = entry.getMetadata().builder();
                      }
                      metadata = builder
-                           .lifespan(configuration.clustering().l1().lifespan())
-                           .build();
+                        .lifespan(configuration.clustering().l1().lifespan())
+                        .build();
                      metadata = new L1Metadata(metadata);
                   }
-               } else {
-                  doCommit = false;
                }
+               isL1Write = true;
+               doCommit = Commit.COMMIT_NON_LOCAL;
             }
 
-            boolean created = false;
-            boolean removed = false;
-            boolean expired = false;
-            if (!isForeignOwned) {
-               created = entry.isCreated();
-               removed = entry.isRemoved();
-               if (removed && entry instanceof MVCCEntry) {
-                  expired = ((MVCCEntry) entry).isExpired();
+            if (doCommit.isCommit()) {
+               boolean created = false;
+               boolean removed = false;
+               boolean expired = false;
+               if (doCommit.isLocal()) {
+                  created = entry.isCreated();
+                  removed = entry.isRemoved();
+                  if (removed && entry instanceof MVCCEntry) {
+                     expired = ((MVCCEntry) entry).isExpired();
+                  }
                }
-            }
 
-            if (doCommit) {
+               // TODO use value from the entry
                InternalCacheEntry previousEntry = dataContainer.peek(entry.getKey());
                Object previousValue = null;
                Metadata previousMetadata = null;
                if (previousEntry != null) {
                   previousValue = previousEntry.getValue();
                   previousMetadata = previousEntry.getMetadata();
                }
-               commitManager.commit(entry, metadata, trackFlag, l1Invalidation);
-               if (!isForeignOwned) {
-                  notifyCommitEntry(created, removed, expired, entry, ctx, command, previousValue, previousMetadata);
+               if (isL1Write && previousEntry != null && !previousEntry.isL1Entry()) {
+                  // don't overwrite non-L1 entry with L1 (e.g. when originator == backup
+                  // and therefore we have two contexts on one node)
```

Comment:
If we have regular writes while we're not an owner, that means `CACHE_MODE_LOCAL` was used, and I think it's safe to say the combination of `CACHE_MODE_LOCAL` and L1 leads to undefined results. Also L1 writes don't acquire a lock, you'd need to do this check in `DataContainer.compute()`.

OTOH I think L1 writes are actually committed in `org.infinispan.interceptors.distribution.L1WriteSynchronizer#runL1UpdateIfPossible()`, so the `COMMIT_NON_LOCAL` branch is only used for L1 invalidations.